content/blog/2022/2022-05-27-one-stop-guide-mapstruct.md

```diff
@@ -0,0 +1,2116 @@
+            }
+        }
+
+        return personDTO;
+    }
+}
+```
+
+In case the adder was not available, the setter would have been used.
+
+### Mapping Streams
+
+Mapping Streams are similar to mapping collections. The only difference is that the auto-generated implementation would return a `Stream` from a provided `Iterable`:
+
+```java
+@Mapper
+public interface CollectionMapper {
+    CollectionMapper INSTANCE = Mappers.getMapper(CollectionMapper.class);
+
+    Set<String> convertStream(Stream<Long> ids);
+
+    @Mapping(source = "degreeName", target = "degree")
+    @Mapping(source = "institute", target = "college")
+    @Mapping(source = "yearOfPassing", target = "passingYear")
+    EducationDTO convert(Education education);
+    List<EducationDTO> convert(Stream<Education> educationStream);
+}
+```
+
+The implementation methods would look something like below:
+
+```java
+@Generated(
+    value = "org.mapstruct.ap.MappingProcessor"
+)
+public class CollectionMapperImpl implements CollectionMapper {
+
+    ...
+
+    @Override
+    public Set<String> convertStream(Stream<Long> ids) {
+        if ( ids == null ) {
+            return null;
+        }
+
+        return ids.map( long1 -> String.valueOf( long1 ) )
+        .collect( Collectors.toCollection( HashSet<String>::new ) );
+    }
+
+
+    @Override
+    public List<EducationDTO> convert(Stream<Education> educationStream) {
+        if ( educationStream == null ) {
+            return null;
+        }
+
+        return educationStream.map( education -> convert( education ) )
+        .collect( Collectors.toCollection( ArrayList<EducationDTO>::new ) );
+    }
+
+    protected EmploymentDTO employmentToEmploymentDTO(Employment employment) {
+        if ( employment == null ) {
+            return null;
+        }
+
+        EmploymentDTOBuilder employmentDTO = EmploymentDTO.builder();
+
+        employmentDTO.designation( employment.getDesignation() );
+        employmentDTO.salary( employment.getSalary() );
+
+        return employmentDTO.build();
+    }
+}
+```
+
+### Mapping Enums
+
+MapStruct allows the conversion of one `Enum` to another `Enum` or `String`. Each constant from the enum at the source is mapped to a constant with the same name in the target. But in the case of different names, we need to annotate `@ValueMapping` with source and target enums.
+
+For example, we will define an enum named `DesignationCode`:
+
+```java
+public enum DesignationCode {
+    CEO,
+    CTO,
+    VP,
+    SM,
+    M,
+    ARCH,
+    SSE,
+    SE,
+    INT
+}
+```
+
+This will be mapped to `DesignationConstant` enum:
+
+```java
+public enum DesignationConstant {
+    CHIEF_EXECUTIVE_OFFICER,
+    CHIEF_TECHNICAL_OFFICER,
+    VICE_PRESIDENT,
+    SENIOR_MANAGER,
+    MANAGER,
+    ARCHITECT,
+    SENIOR_SOFTWARE_ENGINEER,
+    SOFTWARE_ENGINEER,
+    INTERN,
+    OTHERS
+}
+```
+
+Now we can define an Enum mapping in the following way:
+
+```java
+@Mapper
+public interface UserMapper {    
```

Comment:
Again, I'm not sure if we need the whole generated code here, also in the next code example (especially as this and the next one are very similar)

Reply:
I have kept this and removed the other as this shows few of important things like the switc-case instead of normal if-else and the auto-handling in case of `<ANY_REMAINING>` or `NULL`.